### PR TITLE
Initial commit for Search functionality

### PR DESCRIPTION
**Pull request by eswarasai**
##### Description

- Improved the Search functionality on the `/explorer`
- Removed URL params for the search keywords. Moved them to `localStorage`
- Added Tech Stack on the left rails which is connected with the search bar
- Add ability to clear search
- Update Sort design to Dropdown style

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)

UI

##### Testing

Checked the search functionality changes against the production search behaviour

##### Refers/Fixes

Fixes: #412

### CODE REVIEW

**Comment by owocki** [comment]
> Removed URL params for the search keywords. Moved them to localStorage

I kind of liked the URL params for search keywords 

this way i can send you a link to metamasks bounties by just pasting http://localhost:8080/explorer?q=Metamask

File: app/assets/v2/css/search_bar.css [hunk 1]
```diff
@@ -110,7 +122,7 @@
   }
 
   .order_by .sort_option:last-child {
-    margin-right: 0px;
+    margin-right: 0;
```
Comment (thelostone-mc) [inline app/assets/v2/css/search_bar.css]: Is that what we follow ?
Didn't pay attention to this until now

File: app/assets/v2/js/pages/dashboard.js [hunk 2]
```diff
@@ -1,6 +1,16 @@
 /* eslint-disable no-loop-func */
 // helper functions
-var sidebar_keys = [ 'experience_level', 'project_length', 'bounty_type', 'bounty_filter', 'network', 'idx_status' ];
+var technologies = [ '.NET', 'ASP .NET', 'Angular', 'Backbone', 'Bootstrap', 'C', 'C#', 'C++', 'CSS', 'CSS3', 'CoffeeScript', 'Dart', 'Django', 'Drupal', 'DynamoDB', 'ElasticSearch', 'Ember', 'Erlang', 'Express', 'Go', 'Groovy', 'Grunt', 'HTML', 'Hadoop', 'Jasmine', 'Java', 'JavaScript', 'Jekyll', 'Knockout', 'LaTeX', 'Mocha', 'MongoDB', 'MySQL', 'NoSQL', 'Node.js', 'Objective-C', 'Oracle', 'PHP', 'Perl', 'Polymer', 'Postgres', 'Python', 'R', 'Rails', 'React', 'Redis', 'Redux', 'Ruby', 'SASS', 'Scala', 'Sqlite', 'Swift', 'TypeScript', 'Websockets', 'WordPress', 'jQuery' ];
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/dashboard.js]: Break this into multiple lines ? Avoids scrolling :P

File: app/assets/v2/css/search_bar.css (hunk 1 above)
Comment (eswarasai) [reply app/assets/v2/css/search_bar.css]: `px` for 0 is redundant. IntelliJ IDE gives a warning. Hence I remove them as it doesn't make sense to add `px` anyway.

File: app/assets/v2/js/pages/dashboard.js (hunk 2 above)
Comment (eswarasai) [reply app/assets/v2/js/pages/dashboard.js]: Sure. Will do. You got any other better suggestions for the Tech Stack which doesn't involve Backend refactor? 😛

**Comment by thelostone-mc** [comment]
@eswarasai Code looks solid  ^_^
Left a comment there

@PixelantDesign Are we sure we wanna display the tech stack on the left rails ? 
Reason I ask : we've got ton of options already and this list will keep growing !
It'lls end up looking ugly :|

If we really want this : I would we could make our search field better by adding support for comma separated values - 
If I were to type - css, html // I'd get issues of either tech stack

**Comment by eswarasai** [comment]
@mbeacom @thelostone-mc -- This one's good for review as well as merge. Let me know if there are any changes required.

@mbeacom -- Is it possible to deploy this PR on staging so that @PixelantDesign can take a look and provide her feedback? Thanks.

File: app/assets/v2/css/search_bar.css (hunk 1 above)
Comment (thelostone-mc) [reply app/assets/v2/css/search_bar.css]: change IDE :P

**Comment by thelostone-mc** [comment]
The image files! 
@eswarasai could we use the font-awesome icons ?

File: app/retail/templates/shared/head.html [hunk 3]
```diff
@@ -21,7 +21,7 @@
   FontAwesomeConfig = { searchPseudoElements: true };
 </script>
 <script defer src="{% static "v2/js/fontawesome-all.min.js" %}"></script>
-<link rel="stylesheet" href="{% static "v2/css/bootstrap.min.css" %}" crossorigin="anonymous">
+<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/css/bootstrap.min.css" crossorigin="anonymous">
```
Comment (thelostone-mc) [inline app/retail/templates/shared/head.html]: worth updating `bootstrap.min.css` with the latest code instead of the CDN ?

**Comment by eswarasai** [comment]
> could we use the font-awesome icons ?

@thelostone-mc -- I don't think I can leverage FontAwesome as the dropdown is that of jQuery UI and it uses the chevron down icon via `background-image`. Although I did try to use FontAwesome icon via CSS `content` using `:before` looks like that didn't work. Do you want to give it a shot, maybe?

**Comment by thelostone-mc** [comment]
@eswarasai i'll take a stab at it once this gets merged ^_^

File: app/retail/templates/shared/head.html (hunk 3 above)
Comment (thelostone-mc) [reply app/retail/templates/shared/head.html]: @mbeacom thoughts ?

File: app/assets/v2/js/pages/dashboard.js [hunk 4]
```diff
@@ -1,6 +1,22 @@
 /* eslint-disable no-loop-func */
 // helper functions
-var sidebar_keys = [ 'experience_level', 'project_length', 'bounty_type', 'bounty_filter', 'network', 'idx_status' ];
+var technologies = [
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/dashboard.js]: @PixelantDesign @mbeacom  we are hardcoding this for now !! Cause I believe no alternative was given during the offline discussion

File: app/retail/templates/shared/head.html (hunk 3 above)
Comment (thelostone-mc) [reply app/retail/templates/shared/head.html]: @eswarasai could you update that ? :P

**Comment by eswarasai** [comment]
@mbeacom -- Just updated the Bootstrap library to the latest version and also removed it being served from the public CDN

**Comment by PixelantDesign** [comment]
Hey @eswarasai could we try the keywords in a scrollbar so that it doesnt get super long?

**Comment by thelostone-mc** [comment]
@PixelantDesign would that make the left rails a lil harder to use ? 🤔

**Comment by eswarasai** [comment]
@mbeacom -- Not really sure how to get the `stickler-ci` and the `travis-ci` errors resolved. Let me know if there's something missing from my end.

In case we want to use the `Keywords` to be plugged in for the Tech Stack, it needs some refactor in the Backend as the current `Keywords` doesn't determine between Technology/GitHub User/GitHub Org etc.

@PixelantDesign -- I can add a scrollbar just for the Tech Stack but it might be really difficult to figure out what are the applied tech stack filters on the left rails. This really needs a lot of UX improvement in case we want the Tech Stack on the left rails, like having a Search, displaying the applied filters on top with an option to clear etc. Let me know how can I proceed on this.

**Comment by mbeacom** [comment]
@eswarasai Don't worry about loading the `Keywords` - I was planning on adding it following any agreed changes to the FE.

**Comment by eswarasai** [comment]
@mbeacom -- Looking forward in getting this merged as I've got @PixelantDesign approval too 😄

**Pull request merged**